app/scripts/contentscript.js

```diff
@@ -11,32 +11,58 @@ const inpageContent = fs.readFileSync(path.join(__dirname, '..', '..', 'dist', '
 const inpageSuffix = '//# sourceURL=' + extension.extension.getURL('inpage.js') + '\n'
 const inpageBundle = inpageContent + inpageSuffix
 
+
+const inpageWeb3Content = fs.readFileSync(path.join(__dirname, '..', '..', 'dist', 'chrome', 'inpage-web3.js')).toString()
+const inpageWeb3Suffix = '//# sourceURL=' + extension.extension.getURL('inpage-web3.js') + '\n'
+const inpageWeb3Bundle = inpageWeb3Content + inpageWeb3Suffix
+
 // Eventually this streaming injection could be replaced with:
 // https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction
 //
 // But for now that is only Firefox
 // If we create a FireFox-only code path using that API,
 // MetaMask will be much faster loading and performant on Firefox.
 
-if (shouldInjectWeb3()) {
-  setupInjection()
-  setupStreams()
+if (shouldInjectProvider()) {
+  listenForProviderRequest()
+}
+
+/**
+ * Establishes listeners for provider injection requests from the browser context
+ * and for provider approvals and rejections from the background script context
+ */
+function listenForProviderRequest () {
+  window.addEventListener('message', (event) => {
+    if (event.source !== window) { return }
+    if (!event.data || !event.data.type || event.data.type !== 'ETHEREUM_PROVIDER_REQUEST') { return }
+    extension.runtime.sendMessage({
+      action: 'init-provider-request',
+      origin: event.source.origin,
+      web3: event.data.web3,
+    })
+  })
+
+  extension.runtime.onMessage.addListener(({ action, web3 }) => {
+    if (!action || action !== 'approve-provider-request') { return }
+    setupStreams()
+    injectScript(web3 ? inpageWeb3Bundle : inpageBundle)
+    window.postMessage({ type: 'ETHEREUM_PROVIDER_SUCCESS' }, '*')
```

Comment:
This event is triggered even so the `injectScript` function might have failed (the error would be logged to console), is that expected behaviour?